### ISSUE TITLE
Valid digest error in request_storage_access javascript file on Heroku

### ISSUE THREAD

**Issue by dansundy** [Shopify/shopify_app#814]
After doing gem updates I started getting the following error for the `request_storage_access` file in production (a Heroku staging server) only:

```
Failed to find valid digest in the 'integrity' attribute for for resource 'https://[myapp]/assets/shopify_app/request_storage_access-[hash].js' with computed SHA-265 integrity '[different-hash]'. The resource has been blocked.
```

It looks like this error happened because I updated Sprockets to `4.0.0` and may be related to this very old but still open issue in `sprockets-rails`: https://github.com/rails/sprockets-rails/issues/393. It looks like I have fixed the issue temporarily by downgrading Sprockets to `3.7.2` by adding `gem 'sprockets', '~> 3.0'` to my `Gemfile`.

I also noticed [this issue](https://github.com/Shopify/shopify_app/issues/784) which may be related. But I can confirm that my assets were compiling properly when deploying to Heroku. I wasn't getting 404 errors.

As mentioned, it looks like I have resolved this temporarily but thought I'd flag the issue in case others were running into it and because it may come up more often in the future.

Environment info:

Browser: Chrome Version 78.0.3904.70 (Official Build) (64-bit)
OS: MacOS Mojave 10.14.6
Ruby: 2.5.7
Rails: 5.2.3
Shopify App: 11.2.0

**Comment by AfolabiOlaoluwa** [comment]
I am using:
Ruby: 2.6.2
Rails: 6.0.0
Shopify App: config.api_version = "2019-10"

Rails 6 comes with sprockets v4.0.0. So I did as @dansundy suggested by adding `gem 'sprockets', '~> 3.0'` to me Gemfile, then I run `bundle update sprockets` which regressed sprockets to v3.7.0. It didn't work for me still.

**Comment by yu-coder** [comment]
@dansundy I had the same issue but your solution helped me.
Thank you!

**Comment by ziaulrehman40** [comment]
I am on rails 6.0.1, ruby 2.6.5 and downgrading sprockets solved this issue. Though my css is not loading for some reason.

**Comment by AfolabiOlaoluwa** [comment]
> I am on rails 6.0.1, ruby 2.6.5 and downgrading sprockets solved this issue. Though my css is not loading for some reason.

Yes CSS ain't loading

**Comment by ziaulrehman40** [comment]
@AfolabiOlaoluwa I fixed css loading issue on my end, it was missing `stylesheet_pack_tag` see https://gist.github.com/ziaulrehman40/19c87cc73dd75ce6e5d717ace8ddcc48#gistcomment-3099095 for detail. Might be same issue in your case as well.

**Comment by AfolabiOlaoluwa** [comment]
@ziaulrehman40 Adding `stylesheet_link_tag` and `stylesheet_pack_tag` actually result to 500 Error for me on Production. Here is the issue I opened regarding that. https://github.com/Shopify/shopify_app/issues/817

Also, you can check the source code here https://github.com/AfolabiOlaoluwa/ShopifyLoyaltyApp

However, I will look through your fix and how that could be possible with Webpack

**Comment by mizurnix** [comment]
Seems like quite a few developers (including me) are building their first Rails 6 app with shopify_app gem right now, and because of the `request_storage_access` issue, many of us came across the  CSS not being loaded issue right after "solving" `request_storage_access` and some may even think these two are somehow related. I believe that's not the case at all - according to [README](https://github.com/rails/webpacker/blob/master/docs/css.md#css-sass-and-scss):

> In order to have styles load in production, you need to include stylesheet_pack_tag with the same name as the javascript file that imports the styles.

**Comment by AfolabiOlaoluwa** [comment]
@mizurnix you might wanna show how you added yours in Rails 6. Here is my [embedded.html.erb](https://github.com/AfolabiOlaoluwa/ShopifyLoyaltyApp/blob/master/app/views/layouts/embedded_app.html.erb). `stylesheet_pack_tag` just couldn't go in the `if ShopifyApp.use_webpacker?`, the app won't run for me.

**Comment by mizurnix** [comment]
@AfolabiOlaoluwa I think it should work both inside AND outside `ShopifyApp.use_webpacker?` since you are using webpack. I can confirm that in my case both variants work (tested in production as well).

This works:

```erb
    ...
    <%= stylesheet_pack_tag 'application' %>
    <% if ShopifyApp.use_webpacker? %>
      <%= javascript_pack_tag 'application', 'data-turbolinks-track': 'reload' %>
    <% else %>
      <%= javascript_include_tag 'application', "data-turbolinks-track" => true %>
    <% end %>
    ...
```

So does this:

```erb
    ...
    <% if ShopifyApp.use_webpacker? %>
      <%= stylesheet_pack_tag 'application' %>
      <%= javascript_pack_tag 'application', 'data-turbolinks-track': 'reload' %>
    <% else %>
      <%= javascript_include_tag 'application', "data-turbolinks-track" => true %>
    <% end %>
    ...

BTW, what do you mean with "app won't run"?

**Comment by AfolabiOlaoluwa** [comment]
Well, so here is the issue it results to [here](https://github.com/rails/webpacker/issues/2071) if I add:

```
<%= stylesheet_link_tag 'application', media: 'all', 'data-turbolinks-track': 'reload' %>
<%= stylesheet_pack_tag 'application', media: 'all', 'data-turbolinks-track': 'reload' %>
```

**Comment by isaacbowen** [comment]
I'm working around this with `Rails.application.config.assets.version = ''`, in my config/initializers/assets.rb, with sprockets 4 and rails 6. This works becaaauuuse:

https://github.com/rails/sprockets/blob/08fef08562c7a6a13a7c521938e83409a33e2b77/lib/sprockets/loader.rb#L198

We're using digests in asset urls, so (unless I'm missing something) it's totally unnecessary for us to have version as a mechanism for expiring assets. Plus, now our integrity hashes work. :D

**Comment by airhorns** [comment]
I think it might be wise to change the `javascript_include_tag` to not pass the `:integrity` option until the upstream bug is fixed in Sprockets, or else everyone's initial Rails 6 apps are going to be broken with this issue that each dev has to dig up and fix. It's an elective feature that enhances security but a secure broken app is not very useful in comparison to a slightly less secure actually working app if you ask me.

Also FWIW if you are going to apply the `Rails.application.config.assets.version = ''` fix in an initializer, it's worth checking if any other railties are setting that version to be something, because some do and do it after files in `config/initializers` run. `react-rails` for example sets the value at this time, so to override it I added this to my `config/application.rb`:

```ruby
    config.after_initialize do
      Rails.application.config.assets.version = ""
    end
```

which is even gross-er, but hey, at least I can access the application!

**Comment by uurcankaya** [comment]
Issue persists in Chrome 81.0.4044.92 

```
Rails 6
gem 'sprockets', '~> 3.0'
```

**Comment by uurcankaya** [comment]
Cloudflare's auto-minify setting causes this problem. It works after that. 

```
Rails 6
gem 'sprockets', '~> 3.0'
```

**Comment by tiagom87** [comment]
> Cloudflare's auto-minify setting causes this problem. It works after disabling that setting.
> 
> ```
> Rails 6
> gem 'sprockets', '~> 3.0'
> ```

This worked for me

**Comment by paulbaudrier** [comment]
```
    config.after_initialize do
      Rails.application.config.assets.version = ""
    end
```

This doesn't work for me with Shopify_app 13.0.1 and rails 6
@tiagom87  @uurcankaya Have the gem 'sprockets' in gemfile.lock but how to disable the setting ?

**Comment by rafaelfranca** [comment]
This issue should be fixed now with sprockets 4.0.1. Let me know if you still have a problem and I'll reopen.
